Allow building with PublishAot set

agocke opened the pull request:
Fixes https://github.com/dotnet/sdk/issues/28475

The current PublishAot behavior is to attempt to restore the runtime pack for NativeAOT immediately. This doesn't work for osx-arm64, since there's no runtime pack for that RID. This causes even restore to fail, when it should succeed (as should build).

The new behavior is to only attempt to restore the runtime pack when publish is actually being run. An error message is also printed if an unsupported RID is used, although an "escape hatch" property is also provided to skip that error.

agocke committed:
Add unit test

agocke committed:
Only set AotEnabled for ProcessFrameworkReferences when publishing

agocke committed:
Update existing test to run on all platforms

agocke committed:
Add error for unsupported RIDs

LakshanF reviewed (approved):
LGTM! ITs good that users can bypass the rid check in the SDK if needed when ILC packages with new rids get added

nagilson commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets:
```diff
@@ -81,6 +81,7 @@ Copyright (c) .NET Foundation. All rights reserved.
 
     <PropertyGroup>
       <_NuGetRestoreSupported Condition="('$(Language)' == 'C++' and '$(_EnablePackageReferencesInVCProjects)' != 'true')">false</_NuGetRestoreSupported>
+      <_AotEnabled Condition="('$(PublishAot)' == 'true') and ('$(_IsPublishing)' == 'true')">true</_AotEnabled>
```
FYI `_IsPublishing` is only for CLI scenarios ATM, we are discussing if we can get it to work in VS Publishing today. We don't think we can get it to work for `t:/Publish` and plan on evaluating it in a Monday partner sync or in a separate call, I was planning on adding you @agocke

agocke replied:
Recommendation on what I should use instead?

nagilson replied:
I'm sorry to say I don't have any immediate recommendations, I am curious as to how `PublishTrimmed`, single file, aot, etc currently work. The dilemma we're facing right now is detecting if we are publishing early in evaluation seems to be incompatible with MSBuild as it currently exists, because of custom Publish targets that call publish later on, --no-build publishes, etc. It seems like it should be so simple and something we really want; we have been going back and forth on this for a few months now!

agocke commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -82,6 +82,10 @@ Copyright (c) .NET Foundation. All rights reserved.
   </Target>
 
+  <PropertyGroup>
+    <_AotSupportedRids>linux-x64;linux-arm64;linux-musl-x64;linux-musl-arm64;windows-x64;windows-arm64</_AotSupportedRids>
```
That list isn't correct -- it includes the macos RIDs. I need the real list here, as I want to provide an error for mac, not just let it go.

agocke replied:
OK, this logic is getting convoluted, but I think this split is what we want now.

We want the full list in restore, so that restore will try to grab it.

But we want the restricted list here, so that you get a clear error when you're doing something unsupported.

I might be missing something though.

agocke replied:
So, the experience I think we want (based on previous changes in this code) is:

1. PublishAot on Mac = error unsupported.
2. PublishAot on Mac w/ explicit bypass = works.

If that's true, the list of acceptable RIDs in (1) does not include osx, but it does in (2). I don't see how those could be the same list.

AntonLapounov commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -111,10 +115,15 @@ Copyright (c) .NET Foundation. All rights reserved.
                  ResourceName="CompressionInSingleFileRequiresSelfContained" />
+    <NETSdkError Condition="'$(PublishAot)' == 'true' And
+                            '$(_SkipAotSupportedRidCheck)' != 'true' And
+                            $(_AotSupportedRids.Contains('$(RuntimeIdentifier)')) != 'true'"
```
This will allow identifiers like `linux-arm` and arbitrary substrings. You might want to do `Contains(';$(RuntimeIdentifier);')` instead.

agocke commented:
Talking with @sbomer offline, maybe the right answer is just to not fail restore at all. We can give a better error message later.

agocke committed:
Respond to PR comments and stop failing runtimepack restore

agocke commented on src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs:
```diff
@@ -362,11 +362,9 @@ var runtimeRequiredByDeployment
-                    Log.LogError(Strings.AotNoValidRuntimePackageError);
-                    return;
-                }
+                // Try to download the package. If it fails, we don't error out, as failing a restore
```
Yup, I'll adjust the comment.

agocke committed:
Merge remote-tracking branch 'upstream/release/7.0.1xx' into build-aot-app

agocke committed:
Add test that checks if the bypass workds

Also upgrades global.json as the build pulls the BundledVersion file
from the toolset SDK.

agocke committed:
Fixup KnownILCompilerPack RIDs

agocke committed:
Revert global.json change as we're just going to overwrite the RID list anyway

agocke committed:
Fix comment

agocke committed:
Remove duplicate test

agocke commented:
Yup, the decision in this PR is that restore will never fail, even if the eventual publish certainly will.

I think this PR now represents a workable solution. Publish will, by default, check that the RID is supported and error if it is not.

If the check is bypassed, publish will attempt to use a downloaded package (which will be downloaded during restore if we've added it to the list of available packages, or if the user adds a new `KnownILCompilerPack`).

agocke committed:
Windows RID typo

jkotas commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -111,10 +115,15 @@ Copyright (c) .NET Foundation. All rights reserved.
                             '$(SelfContained)' != 'true'"
                  ResourceName="CompressionInSingleFileRequiresSelfContained" />
+    <NETSdkError Condition="'$(PublishAot)' == 'true' And
+                            '$(_SkipAotSupportedRidCheck)' != 'true' And
```
We have similar property in 
https://github.com/dotnet/runtime/blob/cb6fb60dfc253bc13655fafa15686cb864fd0ea2/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets#L58 . 

Would it make sense to use one property to suppress all these supported checks?

- Rename `_SkipAotSupportedRidCheck` -> `SkipAotSupportedCheck` here
- Rename `DisableUnsupportedError` to `SkipAotSupportedCheck` in dotnet/runtime.

agocke replied:
That makes sense, but I'd like to do that as a separate change, as I intend for this to go into .NET 7 as a servicing fix.

jkotas replied:
If you agree that we should use the same property name in all places, the property should be renamed in this servicing fix.

agocke commented:
@sbomer Anything else here?

dsplaisted commented on src/Layout/toolset-tasks/OverrideAndCreateBundledNETCoreAppPackageVersion.cs:
```diff
@@ -123,6 +123,12 @@ void CheckAndReplaceAttribute(XAttribute attribute)
+            // produce, as if don't try to download a missing package then restore will succeed and
+            // we can give a better error during publish that the given RID is not supported.
+            var ilcompilerRids = itemGroup.Elements(ns + "KnownILCompilerPack").First().Attribute("ILCompilerRuntimeIdentifiers");
+            ilcompilerRids.Value = "linux-x64;linux-musl-x64;linux-arm64;linux-musl-arm64;win-x64;win-arm64;osx-x64";
```
Ideally we shouldn't hard-code the supported RuntimeIdentifiers in a C# task.  What's the reason for the mismatch between the values coming from the item and what's actually supported?

agocke replied:
The mismatch is that we were trying to avoid blocking people who wanted to use unsupported, but available packages.

But being too broad here actually causes problems because attempting to download a missing package blocks restore/build.

kant2002 replied:
We may want to have `browser-wasm` here for NativeAOT-LLVM support.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -82,6 +82,10 @@ Copyright (c) .NET Foundation. All rights reserved.
   </Target>
 
+  <PropertyGroup>
+    <_AotSupportedRids>;linux-x64;linux-arm64;linux-musl-x64;linux-musl-arm64;win-x64;win-arm64;</_AotSupportedRids>
```
Usually when we add support for new RuntimeIdentifiers that ends up getting authored in the installer repo and flowing into items such as the KnownFrameworkReference or other items in Microsoft.NETCoreSdk.BundledVersions.props.  Should we do that for this?  How would we support it if we add support for new RuntimeIdentifiers for AoT, and the list of supported RIDs depends on the target framework?

agocke replied:
Note that the list above (which is the supported list that flows from /installer), is not *quite* the same as the supported rids here, namely that osx-x64 is in the "to download" list but not in the "supported" list.

jkotas replied:
The only thing that matters for unsupported architectures is that there is some way for people who want to give it try. It does not need to be pretty. It is ok if it requires extra boilerplate like explicit reference to the osx package.

Does the new SkipAotSupportedCheck property enable that? If yes, I think it would be fine to get rid of this special-casing for available, but officially unsupported architectures.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -111,10 +115,15 @@ Copyright (c) .NET Foundation. All rights reserved.
                  ResourceName="CompressionInSingleFileRequiresSelfContained" />
+    <NETSdkError Condition="'$(PublishAot)' == 'true' And
+                            '$(_SkipAotSupportedRidCheck)' != 'true' And
+                            $(_AotSupportedRids.Contains(';$(NETCoreSdkRuntimeIdentifier);')) != 'true'"
```
Instead of checking whether a string contains the RuntimeIdentifier, shouldn't we be checking to see whether there is a valid `HostILCompilerPack` or `TargetILCompilerPack` instead for this error?
